### PR TITLE
Added ability to use username/password to save the token

### PR DESCRIPTION
**Pull request by wharter**
Added the ability to enter username/password to get and save the token. This was mentioned on our call this morning and I was asked to implement it.

### CODE REVIEW

**Comment by Tahnan** [comment]
To be honest, I'd really like to *not* include this functionality.

* One goal of the auth improvements we made last year was increased security, and to that end, we deliberately changed the UI so that you could ask for an API token and copy it once, but never be able to retrieve it again; honestly I wish we'd removed the endpoint that the v4 client's `save_token` uses to retrieve the user's tokens.

* There was also, in moving from the v4 API to the v5 API, an attempt to have people *never* use their password to connect to the API, and to only use a token; that's why `connect_with_username_and_password` isn't in the documentation, and why its docstring says "PLEASE NOTE: This method is being provided as a temporary measure.  We strongly encourage users of the Luminoso API to use a long-lived token instead, as explained in the V5_README file."

Unless there's a pressing reason that people *must* be allowed to go through the v4 API to get a token instead of going through the UI, this PR strikes me as going against those security improvements.

**Comment by wharter** [comment]
That sounds reasonable. This was after a conversation with @henningsmith where I thought it would be good add a script to se_tools that does this same functionality. It's a common stumbling point and slow down when trying to help people in our API trainings. The save_token.py version without this change does greatly simplify the API token saving. Although it does suffer with the case you just fixed in all the se_tool scripts where the token could/would be saved in the history file and a one time use of the username with a hidden password would seem a little more secure in that regard.

If the issue is that the token could be retrieved multiple times using the v4 api using a username/password then maybe the api should change and only return a temporary password a single time for each call.  There are OEM cases where our customers have expressed that this username/password login is essential.  When we were changing from V4->V5 this was a significant cause for a large customer to walk away when they couldn't get V5 to do what they needed.  I think there were other work arounds they could have chosen, but they had already spent a lot of time struggling just to get the auth working and gave up when they realized they couldn't do what they wanted strictly with the V5 API.

I think the real solution is OAuth where the authentication is decoupled from the token generation, but I'm sure we have better ways to spend resources.

**Comment by henningsmith** [comment]
Yeah and this is different than what I was thinking. I thought if you passed a username and password to this it would use the v5 API to actually just generate a new token and save that, so the end user never sees the token and it is tagged as a new token generated by the lumi-save-token app at a certain date.

Something like
        client = LuminosoClient.connect_with_username_and_password(url=args.domain,username=args.username,password=password)
        token = client.post('/api/v5/tokens/', description='generated by lumi-save-token', password=password)
        client.save_token(token['token'], domain=domain,
                                token_file=args.token_file)

**Comment by wharter** [comment]
Didn't realize there was a way to login with V5 using username and password.  I'll change it to use connect_with_username_and_password, but under the hood it is still using the v4 api and it's just a helper function in the v5 client.

@Tahnan if I implement it with connect_with_username_and_password and get the token with the /token endpoint will that be something you would want?

**Comment by Tahnan** [comment]
> @Tahnan if I implement it with connect_with_username_and_password and get the token with the /token endpoint will that be something you would want?

Honestly, I'd really like to get rid of `connect_with_username_and_password`.  Note that the changelog explicitly describes it as something "to ease the transition from v4 to v5", and that it was added last February.

**Comment by henningsmith** [comment]
OK, Lets kill this PR then.

**Comment by wharter** [comment]
So there should be no way to log into Daylight with the API? I can't think of any system that doesn't have a way to login with a user name and password using an API.  Most use OAuth these days, but many use similar mechanics to what we already have in our token system (if we had a user login first).  Our customers agree with this stance and we lost at least one significant deal because we were not going to support user login even to retrieve a temporary token.

**Comment by Tahnan** [comment]
> I can't think of any system that doesn't have a way to login with a user name and password using an API.

https://docs.github.com/en/rest/overview/other-authentication-methods#via-username-and-password:

> Password authentication to the API will be removed on November 13, 2020.

https://developer.atlassian.com/cloud/jira/platform/deprecation-notice-basic-auth-and-cookie-based-auth/

> Atlassian has introduced support for API tokens for all Atlassian Cloud sites as a replacement for basic authentication requests that previously used a password or primary credential for an Atlassian account, as well as cookie-based authentication.  Basic authentication with passwords and cookie-based authentication are now deprecated and will be removed in 2019 in accordance with the Atlassian REST API policy.

But I can see the use case for "you can connect with a password *if* what you're doing is getting a token so that you'll never have to connect with a password again".  Let me think about what that would look like.

**Review by Tahnan — CHANGES_REQUESTED** [review]

File: luminoso_api/save_token.py [hunk 1]
```diff
@@ -32,8 +39,18 @@ def _main(argv):
...
+    if args.token:
+        LuminosoClient.save_token(args.token, domain=domain,
+                                  token_file=args.token_file)
+
+    if args.username:
+        password = getpass.getpass()
+        client = LuminosoClient.connect_with_username_and_password(url=args.domain,
+                                                                   username=args.username, password=password)
+        token = client.post('/tokens/', description='generated by lumi-save-token',
+                            password=password)
```
[the remark this answers is not in the document]
Comment (Tahnan) [reply luminoso_api/save_token.py:51]: (Not that someone in a SAML environment would even be able to give their password in the first place, I guess, so `connect_with_username_and_password` wouldn't work there either...)

**Comment by Tahnan** [comment]
The more I looked at the changes in `lumi-save-token`, the more they seemed like things we might as well put into `LuminosoClient.save_token()` (thereby leaving the CLI command a wrapper around the method).  I've made a PR, https://github.com/LuminosoInsight/luminoso-api-client-python/pull/69, for that approach, as an alternative to this one.

**Comment by wharter** [comment]
Both GitHub and Atlassian still allow username and password to generate a token, you just have to change to using OAuth to generate that token.

Directly from: The OAuth 2.0 Authorization Framework
[https://tools.ietf.org/html/rfc6749#section-1.3.3](https://tools.ietf.org/html/rfc6749#section-1.3.3)
> 1.3.3.  Resource Owner Password Credentials

>   The resource owner password credentials (i.e., username and password)
   can be used directly as an authorization grant to obtain an access
   token.  The credentials should only be used when there is a high
   degree of trust between the resource owner and the client (e.g., the
   client is part of the device operating system or a highly privileged
   application), and when other authorization grant types are not
   available (such as an authorization code).

>   Even though this grant type requires direct client access to the
   resource owner credentials, the resource owner credentials are used
   for a single request and are exchanged for an access token.  This
   grant type can eliminate the need for the client to store the
   resource owner credentials for future use, by exchanging the
   credentials with a long-lived access token or refresh token.

**Comment by henningsmith** [comment]
@wharter I think we can close out this PR as Lance made https://github.com/LuminosoInsight/luminoso-api-client-python/pull/69 that essentially does the same thing but in a slightly different way.

**Comment by moss** [comment]
@wharter I'm closing this in favor of Lance's PR for the same thing, since that one will also allow password-based login if you're using the API client library directly. Let me know if there's functionality from this one that's missing there -- I don't believe there should be.

I strongly agree that
1) People need to be able to easily authenticate with the API
2) In the long run OAuth is the right way to do that
3) In the short run this (or the equivalent PR #69) will do the trick.

**Pull request closed without merging**